feat(eval): integrate node_modules resolution

ArnavK-09 opened the pull request:
/claim #182

May need more patches later this pr submits as basis for node modules resolution with req tests

ArnavK-09 committed:
base

ArnavK-09 committed:
1

ArnavK-09 committed:
2

ArnavK-09 committed:
3

ArnavK-09 committed:
4

ArnavK-09 committed:
5

ArnavK-09 committed:
Merge branch 'tscircuit:main' into node

ArnavK-09 committed:
Update eval-compiled-js.ts

ArnavK-09 committed:
Update import-eval-path.ts

ArnavK-09 committed:
Update import-node-module.ts

ArnavK-09 committed:
Update import-local-file.ts

ArnavK-09 committed:
Update import-node-module.ts

ArnavK-09 committed:
fmt

seveibar reviewed (commented):
this looks good, but note we want exactly one test per file, it makes it much easier to debug and figure out failing tests. You can just number each file.

Generally looks very good!!!

seveibar commented on lib/utils/resolve-node-module.ts:
```diff
@@ -0,0 +1,148 @@
+  }
+
+  return tryResolveInNodeModules(basePath)
+}
```
The code structure here uses a recursive inline function, instead of that, you should use a `Context` pattern

recursive inline functions are notoriously impossible to refactor. This file will just grow and grow. If you define a nice `Context` interface, you'll find that it's much much easier to split this up into smaller functions

seveibar replied:
Here's an example of a `Context`:

```tsx
interface NodeResolutionContext {
  fsMap: Record<string, string>
}
```

You can then have methods like:

```tsx
export const findPackageJson = (path: string, ctx: NodeResolutionContext) => {
  // ...
}
```

ArnavK-09 committed:
refactor(node-resolution): modularize node module resolution logic

Break down the node module resolution function into smaller, focused helper functions to improve readability and maintainability. This includes creating a context object, separating package.json handling, and isolating export path resolution logic

ArnavK-09 committed:
test(node-resolution): add test for node_modules resolution with tsx files

refactor(webworker): simplify node module import logic by using importLocalFile

The node_modules resolution test ensures that tsx files are correctly resolved and imported from node_modules. The refactor in the webworker module replaces the complex Babel-based transformation with a simpler approach using importLocalFile, improving maintainability and reducing dependencies.

seveibar commented:
@ArnavK-09 this lgtm, think it's ready to merge?

ArnavK-09 commented:
> @ArnavK-09 this lgtm, think it's ready to merge?

yup ig
from my side its working, altough very very very complex node-modules might fail, but this pr performs as basics, since all tests are passing, i dont think this would break anything present already (hope so test coverage is good),

seveibar commented:
/tip $10
